### ISSUE TITLE
xconfig files

### ISSUE THREAD

**Issue by danpovey** [kaldi-asr/kaldi#1124]
@vijayaditya, please note..

I actually have a proposal that will make it easier to integrate things like the skip-layer connections in future, and will make testing and refactoring the nnet3 config generation scripts easier in future.  It's to separate the config-generation into two stages, with an intermediate pseudo-config file [I'm calling this an xconfig file].  Of course it would be fine to have a convenience script that calls the second stage for you.

The proposal is to have a simpler, intermediate version of the nnet3 config files-- kind of like the latex/tex distinction.  The config generator would initially generate the simper version, then would parse it and generate the full config file(s).  The xconfig file would not contain _all_ the information in the full version, just the basic dimension information and the info about how they connect together.  LSTMs and BLSTMs would be treated as if they were components, and so would affine nodes with their nonlinearities.  There is some shorthand here, with input@-3 meaning Offset(input, t, -3).
For example:

```
input-node name=input dim=40
input-node name=ivector dim=100
affine-layer name=affine1 nonlinearity=relu+normalize dim=1024 input=Append(input@-3, input@-2, ... input@3, ivector)
lstm-layer name=lstm2 cell-dim=512 recurrence-offset=-3 projection-dim=256 input=Append(Offset(affine-node, t, -3), affine-node)
blstm-layer name=blstm3 recurrence-offsets=-3,3 cell-dim=512 projection-dim=256 input=lstm2
affine-layer name=final-affine dim=8214 input=blstm3
affine-layer name=final-affine-xent dim=8214 input=blstm3
output-node name=output input=final-affine
output-node name=output-xent input=final-affine-xent
```

Then if we wanted to introduce skip-layers, like what Pegah is doing, it would be something like (giving a simple case):

```
input-node name=input dim=40
affine-layer name=affine1 dim=1024 nonlinearity=relu+renorm input=Append(input@-3, input@-2, ... input@3)
affine-layer name=affine2 dim=1024 nonlinearity=relu+renorm input=Append(affine1@-3, affine1, affine1@3)
affine-layer name=affine2-skip nonlinearity=relu+renorm+dropout dim=1024 input=Append(input@-3, input@-2, ... input@3)
affine-layer name=affine3 nonlinearity=relu+renorm dim=1024 input=Append(affine2@-3, Sum(affine2, affine2-skip), affine2@3)
...
```

To generate the config that removes the skip layers (after dropout-proportion goes to 1), the program that generates the "final" config could just find each component-node where the 'input' expression has things of the form Sum(x, y-skip.zzz), and write out a config file that has those expressions replaced with just x.

All of the generation, parsing and manipulation of these xconfig files would be done in python; the C++ layer would know only about the "real" config files. Also, as I said before, there is no attempt to encode _all_ the information needed to generate the real config file, into the xconfig file.  The python script that goes from the xconfig file to the real config file will take a bunch of options, as needed, which will supply defaults for various layer types.  There may be cases where we want to allow the overriding of those options, and those can be done using name=value things in the xconfig file.  BTW, something we should probably support is to have the a utility that reads an xconfig file, "fills in" all of the implicit defaults and writes it out, so you have an annotated xconfig file with a bunch of extra information in it, including derived information like input-dim and output-dim that will be required in generating the real config file.  This would make debugging much easier.

**Comment by danpovey** [comment]
BTW, I think it would be better to have a totally separate library and new frontend scripts for this, in steps/nnet3/config/.
That will mean there is less need to worry about back compatibility.
If the different types of layer are done as classes, it might make it easier to add new layer types in future.  But I'm not such a great python programmer, so you might have better ideas.

**Comment by danpovey** [comment]
@vijayaditya, I'm not sure when you're coming back from India and whether you have time to work on this?  If you don't think you'll have time soon, I might start it myself.

Another idea:
to enable a compact form of the xconfig file, in the descriptor parsing, something like
input=Append(affine1@-3, affine1@0, affine1@3)
can be abbreviated to just
input=Append(-3, 0, 3)
and if there is no layer specified, it defaults to the previous layer or input-node.  In terms of parsing descriptors, a numeric value x would be interpreted as Offset(previous-layer, x).
Note: affine1@-3 would be interpreted as Offset(affine1, -3), and when the xconfigs were printed out we'd do it in a more verbose Descriptor-like format like Offset(affine1, -3), removing all of these abbreviations.
Note: we'll assume that the xconfigs form a directed acyclic graph (since any cycles should remain within layers), and that they are specified in order, which makes things easier to parse.

My current idea is that there would be a function
 ParseXconfigLine()
that would return an object for each of the lines.  The object would be of a type that would depend on the first field of the line, e.g. 'input-node' -> InputNode, 'affine-layer' -> AffineLayer, 'lda-layer' -> LdaLayer, and so on.
Each object would have methods:
  OutputDim()  [gives the dimension of the output of the layer]
  Name()  [gives the name of the layer, e.g. 'input', 'affine1', etc.
  OutputName()  [gives the name of the output-node of the layer, which might be 'input', or 'affine1.relu' or whatever.]

  PrintXconfigLine(file) [prints the xconfig line for this layer.. this prints it in a normalized form, not necessarily textually the same as the input form, and possibly containing unspecified default parameters.]
  PrintConfigLines(file)  [prints the config-file lines for this layer. ]

Rather than relying on the layer-by-layer discriminative pretraining, which is a hassle to generate config files for, I'd rather support the ephemeral-connection, ResNet-like ideas that Pegah is working on, which seem to work better and also seem to be less of a hassle to work with.  What we could do is have a function

  PrintPrunedConfigLines(file)  [prints any config-file lines for this layer that are to be changed in the final model, possibly removing ephemeral connections.  Will print no lines if this layer does not include ephemeral connections.]


 The xconfig objects should, in future, have some capabilities for editing, that is, we shouldn't be limited to the information specified by the user in the initial xconfig file.  For example, suppose we want to add spatial regularization to the entire network, or add skip-layer ephemeral connections everywhere, we could figure out an easy way to do this via either a program that reads and writes xconfig files, or the program that reads xconfig files and writes regular config files.  These types of manipulations of xconfig files would be available in the library somewhere.  This is not something we would need to support in the first iteration of this stuff-- just bear it in mind when designing things.

The Descriptors [used in fields like input=xxx] should probably be some kind of object that recursively stores other objects, and I think a Python type version of the GeneralDescriptor that you'll see in the C++ code will probably be suitable.  There are certain rules about in what order different operators can appear in descriptors, but we don't need to check these rules in the python level, at least not for now-- just assume that anything goes, like in the GeneralDescriptor.

For parsing the generic format of xconfig lines, which is like '<first-field> then <name>=<value>', where <value> may contain spaces, for now let's just assume that <value> never contains the '=' sign, which makes the parsing easy.  We can extend it later if needed.

**Comment by vijayaditya** [comment]
Addressed in #1197
